### PR TITLE
Pull in retry code.

### PR DESCRIPTION
**Pull request by ruiaraujo**
Simplified version of node-retry with a different internal API to simplify the code and types.

First step in adding an hedged request retry strategy.

### CODE REVIEW

File: lib/retry.ts [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+  return new RetryOperation(timeouts(options), fn);
+}
+
+export function timeouts(options: OperationOptions) {
+  if (options.minTimeout > options.maxTimeout) {
+    throw new Error("minTimeout is greater than maxTimeout");
+  }
+
+  const timeouts: number[] = [];
+  let i;
```
Comment (topicus) [inline lib/retry.ts]: Should the var definition be in the `for`?

File: lib/client.ts [hunk 2]
```diff
@@ -597,13 +602,10 @@ export class ServiceClient {
...
-    const operation = retry.operation(opts);
     const retryErrors: ServiceClientError[] = [];
-
-    const breaker = this.getCircuitBreaker(params);
-
-    return new Promise<ServiceClientResponse>((resolve, reject) =>
-      operation.attempt((currentAttempt: number) => {
+    return new Promise<ServiceClientResponse>((resolve, reject) => {
+      const breaker = this.getCircuitBreaker(params);
+      const retryOperation = operation(opts, (currentAttempt: number) => {
```
Comment (oporkka) [inline lib/client.ts]: I am not exactly sure that the simplifications of this file make it worth it starting to fork and maintain a library as part of Perron.

Comment (ruiaraujo) [reply lib/client.ts]: Updated the description to make it clearer.

**Comment by ruiaraujo** [comment]
:+1:

File: lib/retry.ts [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+  /**
+   * Whether to [unref](https://nodejs.org/api/timers.html#timers_unref) the setTimeout's.
+   * @default false
+   */
+  unref?: boolean;
+  /**
+   * The maximum amount of times to retry the operation.
+   * @default 10
+   */
+  retries: number;
```
Comment (oporkka) [inline lib/retry.ts]: Where are the defaults defined? Some of the numbers look different here
https://github.com/zalando-incubator/perron/blob/800cdf824fb353f08244ced6db5059cc072cf6da/lib/client.ts#L138-L142

Comment (ruiaraujo) [reply lib/retry.ts]: We use different defaults.

Comment (ruiaraujo) [reply lib/retry.ts]: Updated the comments to match our defaults.

**Comment by oporkka** [comment]
:+1:

**Comment by ruiaraujo** [comment]
:+1:

**Pull request merged**